ts/src/profilers/heap-profiler.ts

```diff
@@ -1,7 +1,55 @@
-// Unimplemented stub of HeapProfiler
-// TODO: implement
+/**
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {perftools} from '../../../proto/profile';
+import {serializeHeapProfile} from './profile-serializer';
+
+const profiler = require('bindings')('sampling_heap_profiler');
+
 export class HeapProfiler {
-  constructor() {
-    throw new Error('Heap Profiler is unimplemented');
+  private enabled: boolean;
+
+  /**
+   * @param intervalBytes - average bytes between samples.
+   * @param stackDepth - upper limit on number of frames in stack for sample.
+   */
+  constructor(private intervalBytes: number, private stackDepth: number) {
+    this.enable();
+  }
+
+  /**
+   * Collects a heap profile when heapProfiler is enabled. Otherwise throws
+   * an error.
+   */
+  profile(): perftools.profiles.IProfile {
+    if (!this.enabled) {
+      throw new Error('Heap profiler is not enabled.');
```

Comment:
Do we catch such errors at the top level and make sure they just get logged into the profiler log and do not kill the customer's program?

Note that it's possible that customer starts their program with heap enabled on one machine and with heap disabled on another machine and the current behavior of the server is that the set of profile types is the same among all clients of the same deployment so the one with heap profiling disabled could be asked by the server to collect the heap data and the agent should ignore that logging a warning.

Reply:
Yes -- these errors are caught at the top level and logged.